### PR TITLE
Specify ticks and axis label positions for 3D plots

### PR DESCRIPTION
**Pull request by scottshambaugh**
## PR summary

Closes https://github.com/matplotlib/matplotlib/issues/10767

This allows for specific control of where the ticks and axis labels are positioned for 3D plots. This is robust to rotating the plots.

There is some refactoring work done here to break out drawing different elements into functions to enable the change, and it made me realize how much axis3d needs some work. Would like to get these upper/lower axes independently customizable, but for now this is a step in the right direction even if it doesn't solve everything.

```python
import matplotlib.pyplot as plt
positions = ['default', 'upper', 'lower', 'both', 'none']
fig, axs = plt.subplots(1, 5, subplot_kw={'projection': '3d'})
for ax, pos in zip(axs.flatten(), positions):
    for axis in ax.xaxis, ax.yaxis, ax.zaxis:
        axis.set_label_position(pos)
        axis.set_ticks_position(pos)
    ax.set(xlabel='x', ylabel='y', zlabel='z', title=pos)
```

image

## PR checklist

- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [ ] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-examples-and-tutorials)
- [x] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/coding_guide.html#new-features-and-api-changes)
- [x] Documentation complies with [general](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-docstrings) guidelines

### CODE REVIEW

**Comment by YongYahn** [comment]
It is recommended that when drawing 3D graphics, the default Z-axis is on the left, in line with the diagram drawn by matlab.

**Comment by scottshambaugh** [comment]
Hi @YongYahn, the intent of this PR is to allow you to customize the appearance to do that. However changing the default would change the look of existing plots from past versions, and we [try to avoid](https://matplotlib.org/stable/devel/contributing.html#coding-guidelines) those sorts of backwards-incompatible changes.

File: lib/mpl_toolkits/mplot3d/axis3d.py [hunk 1]
```diff
@@ -320,68 +417,59 @@ def draw_pane(self, renderer):
...
-    @artist.allow_rasterization
-    def draw(self, renderer):
-        self.label._transform = self.axes.transData
-        renderer.open_group("axis3d", gid=self.get_gid())
+    def _axmask(self):
+        info = self._axinfo
+        index = info["i"]
+        axmask = [True, True, True]
+        axmask[index] = False
+        return axmask
```
Comment (oscargus) [inline lib/mpl_toolkits/mplot3d/axis3d.py:425]: I'd suggest either passing `index` as an argument or just stack the calls. I added the explicit variables to reuse the variables, but no reuse here.

File: lib/mpl_toolkits/mplot3d/axis3d.py [hunk 2]
```diff
@@ -445,39 +534,84 @@ def draw(self, renderer):
         self.offsetText.set_ha(align)
         self.offsetText.draw(renderer)
 
-        # Draw ticks:
-        tickdir = self._get_tickdir()
-        tickdelta = deltas[tickdir] if highs[tickdir] else -deltas[tickdir]
+    def _draw_labels(self, renderer, edgep1, edgep2, labeldeltas, centers,
+                     dx, dy):
+        info = self._axinfo
```
Comment (oscargus) [inline lib/mpl_toolkits/mplot3d/axis3d.py:539]: Store `self._axinfo['label']` instead?

File: lib/mpl_toolkits/mplot3d/axis3d.py [hunk 3]
```diff
@@ -320,68 +417,59 @@ def draw_pane(self, renderer):
...
+            tick.tick1line.set_linewidth(tick_lw[tick._major])
+            tick.draw(renderer)
 
-        # Draw Offset text
+    def _draw_offset_text(self, renderer, edgep1, edgep2, labeldeltas, centers,
+                          highs, pep, dx, dy):
+        # Get general axis information:
+        info = self._axinfo
+        index = info["i"]
+        juggled = info["juggled"]
```
Comment (oscargus) [inline lib/mpl_toolkits/mplot3d/axis3d.py:472]: Possibly add `tickdir = info["tickdir"]`?

**Review by oscargus — APPROVED** [review]
Looks good! Nice to have it a bit refactored so that it is easier to see what each part does.

Some comment to slightly improve the code efficiency/readability.

File: lib/mpl_toolkits/mplot3d/axis3d.py [hunk 4]
```diff
@@ -183,6 +186,54 @@ def get_minor_ticks(self, numticks=None):
...
+                           position=position)
+        self._ticks_position = position
+
+    def get_ticks_position(self):
+        """
+        Get the ticks position.
+
+        Returns
+        -------
+        position : {'lower', 'upper', 'both', 'default', 'none'}
```
Comment (QuLogic) [inline lib/mpl_toolkits/mplot3d/axis3d.py:208]: Type only for single-value `Returns`, no name.

File: lib/mpl_toolkits/mplot3d/axis3d.py [hunk 5]
```diff
@@ -183,6 +186,54 @@ def get_minor_ticks(self, numticks=None):
...
+                           position=position)
+        self._label_position = position
+
+    def get_label_position(self):
+        """
+        Get the label position.
+
+        Returns
+        -------
+        position : {'lower', 'upper', 'both', 'default', 'none'}
```
Comment (QuLogic) [inline lib/mpl_toolkits/mplot3d/axis3d.py:232]: Ditto.

File: lib/mpl_toolkits/mplot3d/axis3d.py [hunk 6]
```diff
@@ -256,26 +306,57 @@ def _get_coord_info(self, renderer):
...
+                edgep1s.append(edgep1_l)
+                edgep2s.append(edgep2_l)
+                position.append('lower')
+            if axis_position in ('upper', 'both'):
+                edgep1s.append(edgep1_u)
+                edgep2s.append(edgep2_u)
+                position.append('upper')
+        return edgep1s, edgep2s, position
+
+    def _get_tickdir(self, position=None):
```
Comment (QuLogic) [inline lib/mpl_toolkits/mplot3d/axis3d.py:359]: It's always called with a position, isn't it?
```suggestion
    def _get_tickdir(self, position):
```

File: lib/mpl_toolkits/mplot3d/axis3d.py [hunk 7]
```diff
@@ -446,39 +533,85 @@ def draw(self, renderer):
         self.offsetText.set_ha(align)
         self.offsetText.draw(renderer)
 
-        # Draw ticks:
-        tickdir = self._get_tickdir()
-        tickdelta = deltas[tickdir] if highs[tickdir] else -deltas[tickdir]
+    def _draw_labels(self, renderer, edgep1, edgep2, labeldeltas, centers,
+                     dx, dy):
```
Comment (QuLogic) [inline lib/mpl_toolkits/mplot3d/axis3d.py:537]:
```suggestion
    def _draw_labels(self, renderer, edgep1, edgep2, labeldeltas, centers, dx, dy):
```

File: lib/mpl_toolkits/mplot3d/axis3d.py [hunk 8]
```diff
@@ -446,39 +533,85 @@ def draw(self, renderer):
...
-            lx, ly, lz = proj3d.proj_transform(*pos, self.axes.M)
+        # Calculate offset distances
+        # A rough estimate; points are ambiguous since 3D plots rotate
+        reltoinches = self.figure.dpi_scale_trans.inverted()
+        ax_inches = reltoinches.transform(self.axes.bbox.size)
+        ax_points_estimate = sum(72. * ax_inches)
+        deltas_per_point = 48 / ax_points_estimate
+        default_offset = 21.
+        labeldeltas = (
+            (self.labelpad + default_offset) * deltas_per_point * deltas)
```
Comment (QuLogic) [inline lib/mpl_toolkits/mplot3d/axis3d.py:570]:
```suggestion
        labeldeltas = (self.labelpad + default_offset) * deltas_per_point * deltas
```

File: lib/mpl_toolkits/mplot3d/axis3d.py [hunk 9]
```diff
@@ -446,39 +533,85 @@ def draw(self, renderer):
...
+                                           minmax, maxmin, self._label_position)):
+            # See comments above
+            pep = proj3d._proj_trans_points([edgep1, edgep2], self.axes.M)
+            pep = np.asarray(pep)
+            dx, dy = (self.axes.transAxes.transform([pep[0:2, 1]]) -
+                      self.axes.transAxes.transform([pep[0:2, 0]]))[0]
+
+            # Draw labels
+            self._draw_labels(renderer, edgep1, edgep2, labeldeltas, centers,
+                              dx, dy)
```
Comment (QuLogic) [inline lib/mpl_toolkits/mplot3d/axis3d.py:614]:
```suggestion
            self._draw_labels(renderer, edgep1, edgep2, labeldeltas, centers, dx, dy)
```

**Comment by scottshambaugh** [comment]
Bump on this, would be nice to get in 3.8.0 with one approval already.

**Review by QuLogic — CHANGES_REQUESTED** [review]
Both 'lower' and 'both' do not have any ticks or labels in the what's new example. Also, you've added a test image, but no test?

File: lib/mpl_toolkits/mplot3d/axis3d.py [hunk 10]
```diff
@@ -183,6 +186,54 @@ def get_minor_ticks(self, numticks=None):
                 obj.set_transform(self.axes.transData)
         return ticks
 
+    def set_ticks_position(self, position):
```
[the remark this answers is not in the document]
Comment (scottshambaugh) [reply lib/mpl_toolkits/mplot3d/axis3d.py:189]: It's to match the 2D API

**Comment by scottshambaugh** [comment]
Thanks @QuLogic, it seems like a commit got dropped somewhere. Should be fixed up now.

File: doc/users/next_whats_new/3d_axis_positions.rst [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+
+You can now specify the positions of ticks and axis labels for 3D plots.
+
+.. plot::
+   :include-source:
+
+   import matplotlib.pyplot as plt
+
+   positions = ['default', 'upper', 'lower', 'both', 'none']
+   fig, axs = plt.subplots(1, len(positions), figsize=(20, 4),
```
Comment (QuLogic) [inline doc/users/next_whats_new/3d_axis_positions.rst:12]: The content width in docs is on 7in wide, so a 20in figure will more than 2x squashed down. Can we make this two rows? Or 2x2 like the test image, and just note which one is default in the title.

File: lib/mpl_toolkits/mplot3d/axis3d.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (QuLogic) [reply lib/mpl_toolkits/mplot3d/axis3d.py:189]: Ah, then this also used to inherit the 2D API, which accepted `top`/`bottom`. Of course those did nothing, but did not fail. We should probably still accept those two options and raise a deprecation warning instead.

**Review by QuLogic — COMMENTED** [review]
I see some of the commits are specifically split out, but others are just clean up; do you want to rebase and squash some of them together?

File: lib/mpl_toolkits/mplot3d/axis3d.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (scottshambaugh) [reply lib/mpl_toolkits/mplot3d/axis3d.py:189]: What's the best way to issue that deprecation warning? I'm having trouble finding an example where we do that for a possible value of an argument.

[the remark this answers is not in the document]
Comment (QuLogic) [reply lib/mpl_toolkits/mplot3d/axis3d.py:189]: Something like:
```
if position in ['top', 'bottom']:
    _api.warn_deprecated('3.8', name=f'{position=}', objtype='argument value', ...)
    return
```

File: doc/users/next_whats_new/3d_axis_positions.rst (hunk 11 above)
Comment (scottshambaugh) [reply doc/users/next_whats_new/3d_axis_positions.rst]: Yeah, made this two rows with a (12, 8) figsize. Hard to go much smaller than that without axis labels overlapping eachother.

**Comment by scottshambaugh** [comment]
Squashed a bunch of them, feel free to squash further on merge. Thanks!

**Review by QuLogic — APPROVED** [review]

**Review by tacaswell — APPROVED** [review]

**Pull request merged**